feat: postpone hash verification until read error or mod load

steviegt6 opened the pull request:
### What is the new feature?

`.tmod` file loading as been adjusted to only verify data blob lengths rather than full hash verification. Instead, hash verification has been postponed to mod loading to minimize unnecessary hashing. Hashes will also be taken if the initial `TmodFile::Read` invocation fails.

### Why should this be part of tModLoader?

It significantly speeds up start-up times.

### Are there alternative designs?

A better hashing algorithm and streaming in the background are possible but unnecessary (the former) and terrible (the latter).

steviegt6 committed:
Experiment with postponing hash calculations.

steviegt6 committed:
Remove accidental using directive.

steviegt6 committed:
Check length early on for easy misses.

steviegt6 committed:
Move hash validation to during mod loading.

steviegt6 committed:
Remove now-useless wrapping.

steviegt6 committed:
Updated comment.

steviegt6 commented:
As a side note: I'm personally of the opinion we should remove hashing entirely. Data corruption of this nature is rare and most mods are now sourced from the Steam workshop, which should verify file integrity itself anyway. Many other mod loaders (such as various Minecraft ones) don't hash and face no issues.

It really seems to exist as a remnant of early tModLoader (the entire `.tmod` file format does, for that matter) with little consideration for its continued existence in the current day. For compatibility reasons, we could continue to supply a hash but just skip checking entirely.

steviegt6 committed:
Better error message.

steviegt6 committed:
Not relevant.

steviegt6 committed:
Open tmod file if not already opened.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -65,6 +68,11 @@ internal FileEntry(string name, int offset, int length, int compressedLength, by
 
+	// Starting position of the hashable part of the stream.
+	private long hashStartPos;
+	private bool hashHasBeenTaken;
+	private bool hashMatches;
```
`bool? hashVerified`

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -45,6 +46,8 @@ internal FileEntry(string name, int offset, int length, int compressedLength, by
 
 	private static string Sanitize(string path) => path.Replace('\\', '/');
 
+	private static readonly SHA1 sha1 = SHA1.Create();
```
Does this actually make a difference to performance? We're not multithreading right now, but it's a reduction in encapsulation and potential for future mistakes.

steviegt6 replied:
It does not and I didn't explicitly intend to leave that in there. The multithreading issues are a good catch.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -436,4 +448,40 @@ private void Upgrade()
+	{
+		IDisposable handle = null;
+		if (fileStream == null)
+			handle = Open();
```
No need for the `if` check here. You need to use `using`. The guard statements should also run before trying to get a file handle.

steviegt6 replied:
You're right: when I was glancing at the implementation of Open I misunderstood the `fileStream != null` check.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -436,4 +448,40 @@ private void Upgrade()
+		var oldPos = fileStream.Position;
+		fileStream.Position = hashStartPos;
+		var result = Hash.SequenceEqual(sha1.ComputeHash(fileStream));
+		fileStream.Position = oldPos;
```
No need to put `Position` back, but you should assert that `sharedEntryReadStream == null` just like in `GetStream`. Or just open a new `FileStream`, I don't think there's any real benefit to reusing the existing one for this.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -436,4 +448,40 @@ private void Upgrade()
 		// Read contract fulfilled
 	}
+
+	internal bool VerifyHash()
```
Should probably be something like `VerifyHash() => hashVerified ??= _VerifyHash()` where `_VerifyHash` has the impl and is `private`.

Chicken-Bones reviewed (changes requested):
Overall seems sensible, just some implementation feedback

Chicken-Bones commented:
> As a side note: I'm personally of the opinion we should remove hashing entirely. Data corruption of this nature is rare and most mods are now sourced from the Steam workshop, which should verify file integrity itself anyway. Many other mod loaders (such as various Minecraft ones) don't hash and face no issues.
> 
> It really seems to exist as a remnant of early tModLoader (the entire `.tmod` file format does, for that matter) with little consideration for its continued existence in the current day. For compatibility reasons, we could continue to supply a hash but just skip checking entirely.

Other mod loaders do hash during the data retrieval process. While I agree that data corruption is rare, and the mod would most likely fail to load if it occurred, I don't entirely trust steam our our own server-client mod syncing. In the past hashing was more important because of the signing and trust chain, but I'd like to keep it if the performance impact can be acceptably mitigated.

steviegt6 committed:
Make requested changes.
- Don't bother with caching the SHA1 instance.
- Condense hash booleans into single ternary (`bool?`).
- Simplify caching of verified hash result.
- Don't bother resetting file stream positions.
- Assert `sharedEntryReadStream` is null.

steviegt6 committed:
Remove need for `sharedEntryReadStream` check by just opening the file again in another stream.

steviegt6 committed:
Missed newline.

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -436,4 +445,31 @@ private void Upgrade()
+
+	private void VerifyHashOrThrow(Exception exception)
+	{
+		if (fileStream == null)
+			throw new IOException($"File not open: {path}", exception);
```
No need for this anymore

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/Core/TmodFile.cs:
```diff
@@ -436,4 +445,31 @@ private void Upgrade()
+		if (fileStream == null)
+			throw new IOException($"File not open: {path}", exception);
+
+		// It shouldn't be possible to get this far with an unknown hash starting
+		// position:
+		// - erroneous user code would fail due to no file stream or similar,
+		// - saving, etc. does not use this.
+		Debug.Assert(hashStartPos != 0);
+
+		if (!VerifyHash())
+			throw new Exception(Language.GetTextValue("tModLoader.LoadErrorHashMismatchCorrupted"), exception);
```
May as well inline this, since the method has no other uses. Then you probably don't need the assert.

Chicken-Bones reviewed (changes requested):
Looking good! Just minor cleanups

steviegt6 committed:
Simplify throwing when the hash is invalid.

Chicken-Bones commented:
[2024.12.2.1](https://github.com/tModLoader/tModLoader/releases/tag/v2024.12.2.1) now on preview
